web/src/composebox_typeahead.js

```diff
@@ -1039,30 +1036,25 @@ export function content_typeahead_selected(item, query, input_element, event) {
     return beginning + rest;
 }
 
-export function compose_content_matcher(completing, token) {
+export function compose_content_matcher(completing, item, token) {
```

Comment:
I did it because handling the type of the matcher it returned would be annoying, since matchers can only take items of their specific type. I'd have to create a new set of types of that have e.g. 
```
{
  type: "emoji";
  matcher: (query: string): (emoji: EmojiSuggestion) => boolean;
}
```

and then have `compose_content_matcher` return a union of those types. 

I didn't realize the performance implications, though. If you think it's worth it to type it the way I described instead (or have another idea for typing the return type of `compose_content_matcher`) I can do that.

Reply:
Given [#frontend > remove_diacritics performance](https://chat.zulip.org/#narrow/stream/6-frontend/topic/remove_diacritics.20performance/near/1689797), yes, we care about the optimization of preprocessing the token once. I pointed this out in https://github.com/zulip/zulip/pull/29200#issuecomment-1979641255.

But ginormous union types and the `@typescript-eslint/consistent-type-assertions` error indicate that the design is wrong. I think the simplest fix is to move the `filter_and_sort_candidates` call from `get_sorted_filtered_items` to within each branch of `get_candidates`; then each branch knows exactly which matcher to use and can pass it directly as an argument, so we don’t need `compose_content_matcher` at all.

Reply:
Ah that works great, thanks!